### PR TITLE
Generic/OpeningFunctionBraceKernighanRichie: improve test coverage

### PR DESCRIPTION
**Pull request by rodrigoprimo**
# Description

This PR improves code coverage for the `Generic.Functions.OpeningFunctionBraceKernighanRichie` sniff and makes some tiny improvements to the sniff code. No user-facing changes were made.

## Related issues/external references

Part of https://github.com/PHPCSStandards/PHP_CodeSniffer/issues/146

## Types of changes

- [ ] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement

## PR checklist

- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] \[Required for new sniffs\] I have added XML documentation for the sniff.

### CODE REVIEW

File: src/Standards/Generic/Tests/Functions/OpeningFunctionBraceKernighanRitchieUnitTest.1.inc [hunk 1]
```diff
@@ -212,3 +212,14 @@ function myFunction($a, $lot, $of, $params)
...
+// phpcs:set Generic.Functions.OpeningFunctionBraceKernighanRitchie checkFunctions 0
+function shouldBeIgnored()
+{}
+// phpcs:set Generic.Functions.OpeningFunctionBraceKernighanRitchie checkFunctions 1
+
+function dnfReturnType(): (Response&SuccessResponse)|AnotherResponse|string
+{}
+
+function contentAfterOpeningBrace() { // Some comment.
+}
```
Comment (jrfnl) [inline src/Standards/Generic/Tests/Functions/OpeningFunctionBraceKernighanRitchieUnitTest.1.inc:225]: What about adding a second variation of this test with non-comment content after the brace ?

File: src/Standards/Generic/Tests/Functions/OpeningFunctionBraceKernighanRitchieUnitTest.3.inc [hunk 2]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+// Intentional parse error (missing opening parenthesis).
+// This should be the only test in this file.
+// Testing that the sniff is *not* triggered.
+
+function missingOpeningParenthesis
```
Comment (jrfnl) [inline src/Standards/Generic/Tests/Functions/OpeningFunctionBraceKernighanRitchieUnitTest.3.inc:7]: What is this testing ? The sniff is not looking at the open parenthesis, but at the opening curly.

**Review by jrfnl — CHANGES_REQUESTED** [review]
@rodrigoprimo Thanks for this PR. I think it's a great start, but can be improved further.


Looking at the **tests**, I see the following:

**Missing**:
* Test with an abstract method (no scope opener).
* Live coding test missing the opening curly.
* Live coding test with an open curly, but missing the close curly x 3 - one with correct spacing, one with too much space before the curly, one with no space before the curly. Both these last two should be fixable (and fixed).

**Superfluous**:
* Live coding test missing open parentheses. The sniff is not looking at that.


Looking at the **handling of tabs**... I don't think the multi-tab situation is handled correctly.

If you run the sniff against test case file 2, you will get the following error for the multi-tab line:
```
 10 | ERROR | [x] Expected 1 space before opening brace; found 3 (Generic.Functions.OpeningFunctionBraceKernighanRitchie.SpaceBeforeBrace)
```

... which has no indication that this is 3 tabs, not 3 spaces.

If tabs are called out separately anyway (as is done in this sniff), let's do it correctly and consistently.

And if we're talking **redundant code**, I think the whole code block from line 75 to line 82 can be removed as long as the `$closeBracket` reference on line 85 is replaced with `null`.

What do you think ?

File: src/Standards/Generic/Tests/Functions/OpeningFunctionBraceKernighanRitchieUnitTest.1.inc (hunk 1 above)
Comment (rodrigoprimo) [reply src/Standards/Generic/Tests/Functions/OpeningFunctionBraceKernighanRitchieUnitTest.1.inc]: :+1:

**Comment by rodrigoprimo** [comment]
Thanks for your detailed review, @jrfnl!

> Test with an abstract method (no scope opener).

There is an abstract method declared in an interface, and I assumed it was enough as the sniff doesn't really check where the method is defined. That being said, having an abstract method declared in a class is not a problem, so I went ahead and added one.

> Live coding test missing the opening curly.

:+1: 

> Live coding test with an open curly, but missing the close curly x 3 - one with correct spacing, one with too much space before the curly, one with no space before the curly. Both these last two should be fixable (and fixed).

I might be missing something, but it seems to me the sniff will bail early in those three cases, as when there is no closing curly brace, `scope_opener` is not set.

This is one of the code examples that I have in mind reading your comment:

```php
function missingClosingCurlyBrace()     {
```

I'm assuming you are not suggesting that I change the behavior of the sniff to handle those cases.

> Live coding test missing open parentheses. The sniff is not looking at that.

I replaced this test with a test for the missing opening curly brace.

> Looking at the handling of tabs... I don't think the multi-tab situation is handled correctly.
> If you run the sniff against test case file 2, you will get the following error for the multi-tab line:
> 10 | ERROR | [x] Expected 1 space before opening brace; found 3 (Generic.Functions.OpeningFunctionBraceKernighanRitchie.SpaceBeforeBrace)
> ... which has no indication that this is 3 tabs, not 3 spaces.
> If tabs are called out separately anyway (as is done in this sniff), let's do it correctly and consistently.

I agree, but I don't see an easy and straightforward way to fix it. Especially as there could be a mix of spaces and tabs used. Would it be ok to open an issue for this instead of fixing it in this PR?

> And if we're talking redundant code, I think the whole code block from line 75 to line 82 can be removed as long as the $closeBracket reference on line 85 is replaced with null.
> What do you think ?

Good catch! I removed it in a new commit.

**Comment by jrfnl** [comment]
> > Live coding test with an open curly, but missing the close curly x 3 - one with correct spacing, one with too much space before the curly, one with no space before the curly. Both these last two should be fixable (and fixed).
> 
> I might be missing something, but it seems to me the sniff will bail early in those three cases, as when there is no closing curly brace, `scope_opener` is not set.
> 
> This is one of the code examples that I have in mind reading your comment:
> 
> ```
> function missingClosingCurlyBrace()     {
> ```
> 
> I'm assuming you are not suggesting that I change the behavior of the sniff to handle those cases.

Well, sort of. I mean, this sniff is limiting its own ability by checking for the `scope_opener` instead of checking for a `T_OPEN_CURLY_BRACKET` after the function signature.

Having said that, I agree that should probably be left as "future scope" (and low priority).


> > Looking at the handling of tabs... I don't think the multi-tab situation is handled correctly.
> > If you run the sniff against test case file 2, you will get the following error for the multi-tab line:
> > 10 | ERROR | [x] Expected 1 space before opening brace; found 3 (Generic.Functions.OpeningFunctionBraceKernighanRitchie.SpaceBeforeBrace)
> > ... which has no indication that this is 3 tabs, not 3 spaces.
> > If tabs are called out separately anyway (as is done in this sniff), let's do it correctly and consistently.
> 
> I agree, but I don't see an easy and straightforward way to fix it. Especially as there could be a mix of spaces and tabs used. Would it be ok to open an issue for this instead of fixing it in this PR?

I do see a straight-forward way to fix this, so I will submit a follow-up PR after merging this one. No need for the issue.

**Review by jrfnl — APPROVED** [review]
Thanks for those updating @rodrigoprimo.

I'm going to rebase the PR and squash a few commits together. Merging once the build has passed.

**Pull request merged**